pysb/simulator/opencl_ssa.py

```diff
@@ -0,0 +1,357 @@
+from __future__ import print_function
+from pysb.simulator.base import SimulationResult
+from pysb.bng import generate_equations
+from pysb.simulator.ssa_base import SSABase
+import numpy as np
+import os
+import random
+import time
+from concurrent.futures import ThreadPoolExecutor
+from functools import partial
+
+
+
+
+try:
+    import pyopencl as cl
+    from pyopencl import array as ocl_array
+    from pyopencl import device_type
+    from mako.template import Template
+except ImportError:
+    cl = None
+    device_type = None
+    ocl_array = None
+
+
+class OpenCLSSASimulator(SSABase):
+    """
+    OpenCL SSA simulator
+
+    This simulator is capable of using either a GPU or multi-core CPU.
+    The simulator will detect and ask which device you would like to use.
+    Alteratively, you can set the device using with an environment variable
+     `PYOPENCL_CTX`
+
+    Requires `OpenCL`_ and `PyPpenCL`_.
+
+    .. _OpenCL :
+        https://www.khronos.org/opencl/
+    .. _PyOpenCL :
+        https://documen.tician.de/pyopencl/
+
+
+    Parameters
+    ----------
+    model : pysb.Model
+        Model to simulate.
+    tspan : vector-like, optional
+        Time values over which to simulate. The first and last values define
+        the time range. Returned trajectories are sampled at every value unless
+        the simulation is interrupted for some reason, e.g., due to
+        satisfaction
+        of a logical stopping criterion (see 'tout' below).
```

Comment:
no tout below, also don't see anything about a logical stopping - is this referring to an earlier version of the code?

Reply:
This is OK, it's a verbatim copy of the base class docstring. tout is part of the SimulationResult class. I agree this is a bit obscure, but it's not something that needs to be addressed in this PR.